### PR TITLE
Kevin/refactor models

### PR DESCRIPTION
**Pull request by kevinchan159**
Addresses #41 
- Refactoring our models so that client side will require less logic.
- Basically adds `travelDistance, startName, endName, totalDuration, routeSummary` fields to every `Route` object.
- Since we are only adding fields, this shouldn't break any old logic on client

### CODE REVIEW

File: src/routers/v1/MultiRouteRouter.js [hunk 1]
```diff
@@ -35,8 +35,10 @@ class MultiRouteRouter extends ApplicationRouter<Array<Object>> {
       return [];
     }
 
+    const originName = 'Current Location';
```
Comment (young-k) [inline src/routers/v1/MultiRouteRouter.js]: why not define this up above, and shouldn't it be all caps?

File: src/utils/ParseRouteUtils.js [hunk 2]
```diff
@@ -8,6 +8,10 @@ import LogUtils from './LogUtils';
 import RealtimeFeedUtils from './RealtimeFeedUtils';
 import RequestUtils from './RequestUtils';
 
+const DIRECTION_TYPE = {
```
Comment (young-k) [inline src/utils/ParseRouteUtils.js]: i like this change :very-nice-kevin:

File: src/utils/ParseRouteUtils.js [hunk 3]
```diff
@@ -261,6 +265,49 @@ function generateBoundingBox(path: Object): Object {
...
 
+/**
+ * Returns the distance between [startCoords] and [endCoords] in miles
+ * @param startCoords
+ * @param endCoords
+ * @returns {number}
+ */
+function getDistanceInMiles(startCoords: Object, endCoords: Object): number {
+  const lat1 = startCoords.lat;
+  const lon1 = startCoords.long;
```
Comment (young-k) [inline src/utils/ParseRouteUtils.js]: prefer `startLat` and `startLong`, `lon1` is confusing

File: src/utils/ParseRouteUtils.js [hunk 4]
```diff
@@ -261,6 +265,49 @@ function generateBoundingBox(path: Object): Object {
...
+  const radlat2 = Math.PI * lat2 / 180;
+  const theta = lon1 - lon2;
+  const radtheta = Math.PI * theta / 180;
+  let dist = Math.sin(radlat1) * Math.sin(radlat2) + Math.cos(radlat1) * Math.cos(radlat2) * Math.cos(radtheta);
+  if (dist > 1) {
+    dist = 1;
+  }
+  dist = Math.acos(dist);
+  dist = dist * 180 / Math.PI;
+  dist = dist * 60 * 1.1515;
```
Comment (young-k) [inline src/utils/ParseRouteUtils.js]: 1.1515 needs a comment, and prefer all of these to be on one line with an explanation

File: src/utils/ParseRouteUtils.js [hunk 5]
```diff
@@ -261,6 +265,49 @@ function generateBoundingBox(path: Object): Object {
...
+  return dist;
```
Comment (young-k) [inline src/utils/ParseRouteUtils.js]: in general this function needs more comments

File: src/utils/ParseRouteUtils.js [hunk 6]
```diff
@@ -261,6 +265,49 @@ function generateBoundingBox(path: Object): Object {
...
+    return 0;
+  }
+
+  const radlat1 = Math.PI * lat1 / 180;
+  const radlat2 = Math.PI * lat2 / 180;
+  const theta = lon1 - lon2;
+  const radtheta = Math.PI * theta / 180;
+  let dist = Math.sin(radlat1) * Math.sin(radlat2) + Math.cos(radlat1) * Math.cos(radlat2) * Math.cos(radtheta);
+  if (dist > 1) {
+    dist = 1;
```
Comment (young-k) [inline src/utils/ParseRouteUtils.js]: why?

File: src/utils/ParseRouteUtils.js [hunk 7]
```diff
@@ -261,6 +265,49 @@ function generateBoundingBox(path: Object): Object {
...
+ * Returns the difference between [departureTime] and [arrivalTime] in minutes.
+ * @param departureTime
+ * @param arrivalTime
+ * @returns {number}
+ */
+function getDifferenceInMinutes(departureTime: string, arrivalTime: string): number {
+  const departureDate = new Date(departureTime);
+  const arrivalDate = new Date(arrivalTime);
+  const diffInMs = arrivalDate - departureDate;
+  const diffInMins = Math.round(((diffInMs % ONE_HOUR_IN_MS) % ONE_HOUR_IN_MS) / ONE_MIN_IN_MS);
```
Comment (young-k) [inline src/utils/ParseRouteUtils.js]: why do we need to take the `%` here? add comments!

File: src/utils/ParseRouteUtils.js [hunk 8]
```diff
@@ -271,7 +318,13 @@ function convertMillisecondsToISOString(dateInMs: number): string {
   return `${new Date(dateInMs).toISOString().split('.')[0]}Z`;
 }
 
-function parseWalkingRoute(data: any, dateMs: number, destinationName: string, isArriveBy: boolean): Object {
+function parseWalkingRoute(
+  data: any,
```
Comment (young-k) [inline src/utils/ParseRouteUtils.js]: prefer not having `any` as a type

File: src/utils/ParseRouteUtils.js [hunk 9]
```diff
@@ -512,14 +587,53 @@ function parseBusRoutes(busRoutes: Array<Object>, destinationName: string): Prom
...
+          // If the direction type is walking, the stopName should be the name of the stop we just got off of.
+          stopName = previousStopName;
+        }
+        return {
+          direction,
+          stayOnBusForTransfer,
+          stopName,
+        };
+      }).filter(summary => summary !== null);
+      // The last element in routeSummary should always be an object where stopName is the destination name.
```
Comment (young-k) [inline src/utils/ParseRouteUtils.js]: can we add a newline above this line just to make the separation more clear

File: src/utils/ParseRouteUtils.js [hunk 10]
```diff
@@ -512,14 +587,53 @@ function parseBusRoutes(busRoutes: Array<Object>, destinationName: string): Prom
...
+      const endName = directions.length > 0 ? directions[directions.length - 1].name : null;
+
+      let previousStopName;
+      const routeSummary = directions.map((dir, index) => {
+        const { name, type, stayOnBusForTransfer } = dir;
+        // We want to ignore the intial walking direction
+        if (index === 0 && type === DIRECTION_TYPE.WALK) {
+          return null;
+        }
+        const direction = {
```
Comment (young-k) [inline src/utils/ParseRouteUtils.js]: define `direction` on one line

File: src/utils/ParseRouteUtils.js [hunk 11]
```diff
@@ -512,14 +587,53 @@ function parseBusRoutes(busRoutes: Array<Object>, destinationName: string): Prom
...
 
+      const startName = directions.length > 0 ? directions[0].name : null;
+      const endName = directions.length > 0 ? directions[directions.length - 1].name : null;
+
+      let previousStopName;
+      const routeSummary = directions.map((dir, index) => {
+        const { name, type, stayOnBusForTransfer } = dir;
+        // We want to ignore the intial walking direction
+        if (index === 0 && type === DIRECTION_TYPE.WALK) {
+          return null;
```
Comment (young-k) [inline src/utils/ParseRouteUtils.js]: prefer putting return statement on the above line

File: src/utils/ParseRouteUtils.js [hunk 12]
```diff
@@ -512,14 +587,53 @@ function parseBusRoutes(busRoutes: Array<Object>, destinationName: string): Prom
...
+        const direction = {
+          busNumber: null,
+          type,
+        };
+        let stopName = name;
+        if (type === DIRECTION_TYPE.DEPART) {
+          direction.busNumber = dir.routeNumber;
+          previousStopName = dir.stops[dir.stops.length - 1].name;
+        } else {
+          // If the direction type is walking, the stopName should be the name of the stop we just got off of.
```
Comment (young-k) [inline src/utils/ParseRouteUtils.js]: prefer this comment on the previous line:

`If walking direction type, stopName should be the name of the stop we just got off`

File: src/utils/ParseRouteUtils.js (hunk 5 above)
Comment (kevinchan159) [reply src/utils/ParseRouteUtils.js]: i copied this function from the interweb so im not too sure what the numbers mean either

File: src/utils/ParseRouteUtils.js [hunk 13]
```diff
@@ -512,14 +587,53 @@ function parseBusRoutes(busRoutes: Array<Object>, destinationName: string): Prom
         };
       }));
 
+      const startName = directions.length > 0 ? directions[0].name : null;
+      const endName = directions.length > 0 ? directions[directions.length - 1].name : null;
+
+      let previousStopName;
+      const routeSummary = directions.map((dir, index) => {
```
Comment (young-k) [inline src/utils/ParseRouteUtils.js]: prefer `direction, index`

File: src/utils/ParseRouteUtils.js (hunk 10 above)
Comment (young-k) [reply src/utils/ParseRouteUtils.js]: this variable also seems to be named improperly, we're looping over the directions already, so isn't this not a `direction` object?

File: src/utils/ParseRouteUtils.js (hunk 5 above)
Comment (young-k) [reply src/utils/ParseRouteUtils.js]: not a good reason, and you should just link the relevant source if necessary

File: src/utils/ParseRouteUtils.js (hunk 10 above)
Comment (kevinchan159) [reply src/utils/ParseRouteUtils.js]: it is still a direction object. this represents a `RouteSummary` object where `direction` is a simplified version of our current directions

File: src/utils/ParseRouteUtils.js (hunk 5 above)
Comment (alanna-zhou) [reply src/utils/ParseRouteUtils.js]: oof

File: src/utils/ParseRouteUtils.js [hunk 14]
```diff
@@ -304,17 +359,35 @@ function parseWalkingRoute(data: any, dateMs: number, destinationName: string, i
...
+      },
+    ];
+
     return adjustRouteTimesIfNecessary({
       arrivalTime,
       boundingBox,
       departureTime,
       directions: [direction],
       endCoords,
+      endName: destinationName,
```
Comment (alanna-zhou) [inline src/utils/ParseRouteUtils.js]: why are these different tho

File: src/utils/ParseRouteUtils.js [hunk 15]
```diff
@@ -304,17 +359,35 @@ function parseWalkingRoute(data: any, dateMs: number, destinationName: string, i
...
       numberOfTransfers: 0,
+      routeSummary,
       startCoords,
+      startName: originName,
```
Comment (alanna-zhou) [inline src/utils/ParseRouteUtils.js]: same here

File: src/utils/ParseRouteUtils.js (hunk 14 above)
Comment (kevinchan159) [reply src/utils/ParseRouteUtils.js]: its what was requested by iOS but you're right, we can prob just change it

File: src/utils/ParseRouteUtils.js [hunk 16]
```diff
@@ -512,14 +589,52 @@ function parseBusRoutes(busRoutes: Array<Object>, destinationName: string): Prom
...
+          stopName = previousStopName;
+        }
+        return {
+          direction: routeSummaryDirection,
+          stayOnBusForTransfer,
+          stopName,
+        };
+      }).filter(summary => summary !== null);
+
+      // The last element in RouteSummaryElement object should always be an object where stopName
```
Comment (young-k) [inline src/utils/ParseRouteUtils.js]: prefer this comment be on one line, don't use words like `should`, be concise

File: src/utils/ParseRouteUtils.js [hunk 17]
```diff
@@ -512,14 +589,52 @@ function parseBusRoutes(busRoutes: Array<Object>, destinationName: string): Prom
...
+        // We want to ignore the intial walking direction
+        if (index === 0 && type === DIRECTION_TYPE.WALK) return null;
+        const routeSummaryDirection = { busNumber: null, type };
+        let stopName = name;
+        if (type === DIRECTION_TYPE.DEPART) {
+          routeSummaryDirection.busNumber = direction.routeNumber;
+          previousStopName = direction.stops[direction.stops.length - 1].name;
+        } else { // If walking direction type, stopName should be the name of the stop we just got off
+          stopName = previousStopName;
+        }
```
Comment (young-k) [inline src/utils/ParseRouteUtils.js]: add more whitespace, it's cluttered and tricky to read!

File: src/utils/ParseRouteUtils.js (hunk 14 above)
Comment (kevinchan159) [reply src/utils/ParseRouteUtils.js]: actually wait no its b/c we have `startCoords` and `endCoords` so its more consistent to have `startName` and `endName`

File: src/utils/ParseRouteUtils.js [hunk 18]
```diff
@@ -512,14 +590,52 @@ function parseBusRoutes(busRoutes: Array<Object>, destinationName: string): Prom
...
+
+        return {
+          direction: routeSummaryDirection,
+          stayOnBusForTransfer,
+          stopName,
+        };
+      }).filter(summary => summary !== null);
```
Comment (meganle) [inline src/utils/ParseRouteUtils.js]: You can do `.filter(Boolean)`

File: src/utils/RouteUtils.js [hunk 19]
```diff
@@ -129,6 +129,7 @@ async function createFinalBusRoutes(
  * @returns {Promise<Object>}
  */
 async function getParsedWalkingAndBusRoutes(
+  originName: string,
```
Comment (yuna878) [inline src/utils/RouteUtils.js]: Update docs?

File: src/utils/ParseRouteUtils.js [hunk 20]
```diff
@@ -271,7 +320,13 @@ function convertMillisecondsToISOString(dateInMs: number): string {
   return `${new Date(dateInMs).toISOString().split('.')[0]}Z`;
 }
 
-function parseWalkingRoute(data: any, dateMs: number, destinationName: string, isArriveBy: boolean): Object {
+function parseWalkingRoute(
```
Comment (yuna878) [inline src/utils/ParseRouteUtils.js]: docs?

**Pull request merged**